### PR TITLE
Update ipns validator

### PR DESCRIPTION
**Pull request by dirkmc**
Update go-ipfs to use new validation function format in go-libp2p-record
https://github.com/ipfs/go-ipfs/pull/4566 was getting messy so I closed and opened this new PR

### CODE REVIEW

File: namesys/publisher.go [hunk 1]
```diff
@@ -295,12 +304,31 @@ func selectRecord(recs []*pb.IpnsEntry, vals [][]byte) (int, error) {
...
+	err := proto.Unmarshal(r.Value, entry)
 	if err != nil {
 		return err
 	}
+
+	// Note: The DHT will actually check the signature so we don't
+	// need to do that here
+
+	// Author in key must match author in record
+	parts := strings.Split(r.Key, "/")
```
Comment (vyzo) [inline namesys/publisher.go]: do we need this split?

Comment (dirkmc) [reply namesys/publisher.go]: You're right the relative path component of the IPNS path is not stored in the DHT ([only in DNS](https://github.com/ipfs/go-ipfs/blob/master/namesys/dns_test.go#L104))

**Comment by vyzo** [comment]
You need to add the new dependencies to package.json using `gx update`.

**Comment by vyzo** [comment]
you also need to `go fmt` -- codeclimate rightfully complains.

**Comment by dirkmc** [comment]
I wanted to ask you guys about the dependencies. I updated the go-libp2p-record dependency across the project, and it warned me that I would also need to update go-libp2p-kad-dht, but when I did that it broke a lot of things because of mismatches between the library versions that the new go-libp2p-kad-dht lib imports vs the library versions this project imports.

Is there some easy way of doing all this package management stuff so it will figure out what needs to be updated and do it one step?

**Comment by vyzo** [comment]
Sounds like some dependency will need to propagage -- @Stebalien usually handles those.

**Comment by whyrusleeping** [comment]
We need to at some point write a guide for "propogating gx dependencies"

**Comment by dirkmc** [comment]
The gold standard is whether a recovering java dev can understand it

**Comment by Stebalien** [comment]
@dirkmc Actually, gx is very Java-esque in it's paranoia. But yes, this is blocked on the cmds PR #4568.

File: namesys/publisher.go [hunk 2]
```diff
@@ -295,12 +303,30 @@ func selectRecord(recs []*pb.IpnsEntry, vals [][]byte) (int, error) {
...
+
+	// Note: The DHT will actually check the signature so we don't
+	// need to do that here
+
+	// Author in key must match author in record
+	pid, err := peer.IDFromString(r.Key)
+	if err != nil {
+		return ErrInvalidAuthor
+	}
+	if string(pid) != string(r.Author) {
```
Comment (Stebalien) [inline namesys/publisher.go]: Do we actually need to call `string` here?

Comment (dirkmc) [reply namesys/publisher.go]: No, we should be able to get rid of that cast once the go-libp2p-peer lib is updated

Comment (Stebalien) [reply namesys/publisher.go]: Got it.

**Comment by dirkmc** [comment]
Once the cmds merge is complete I will update the deps and test manually

**Comment by Stebalien** [comment]
So, unfortunately, this actually breaks things (the republisher test cases) because we *allow* publishing IPNS records using one key and then signing the DHT record with another. For example, when publishing an IPNS record with an alternative key, we sign it with the alternative key and then sign the DHT record with the peer key.

The correct solution here is to:

1. Get rid of the DHT signature/author field. We can actually do this without breaking *anything*.
2. Validate the IPNS (not DHT) signature in the validator.

Unfortunately, validating the IPNS signature in the validator is a bit tricky... so I'm not going to fix that in my existing pr (#4610).

**Comment by dirkmc** [comment]
That makes sense, if there's a peer ID in the key there's no need to replicate it in the DHT record.
So are you imagining that the DHT would create a `map[peer.ID]PubKey` from the peer store and pass that to the validator function?

**Comment by Stebalien** [comment]
> So are you imagining that the DHT would create a map[peer.ID]PubKey from the peer store and pass that to the validator function?

No, we'll probably construct the IPNS validator as a closure and have it capture the peerstore (so it can extract the keys from that).

Ideally, the validator would take some additional "validation" information but we'll probably need to wait for a general-purpose IPRS for that.

**Comment by dirkmc** [comment]
Makes sense. Should we close this PR now that it's superseded by #4613?

**Comment by Stebalien** [comment]
Yes.

**Pull request closed without merging**